src/connection/closing.rs

```diff
@@ -0,0 +1,117 @@
+use crate::connection::Result;
+use crate::connection::{ControlCommand, StreamCommand};
+use crate::frame::Frame;
+use crate::{frame, ConnectionError};
+use futures::channel::mpsc;
+use futures::stream::Fuse;
+use futures::{ready, AsyncRead, AsyncWrite, SinkExt, StreamExt};
+use std::collections::VecDeque;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+/// A [`Future`] that gracefully closes the yamux connection.
+#[must_use]
+pub struct Closing<T> {
+    state: State,
+    control_receiver: mpsc::Receiver<ControlCommand>,
+    stream_receiver: mpsc::Receiver<StreamCommand>,
+    pending_frames: VecDeque<Frame<()>>,
+    socket: Fuse<frame::Io<T>>,
+}
```

Comment:
Question not suggestion: Why deliberately implement this as a state machine instead of a procedural `async/await`?

Reply:
So that it can be named without boxing it up.

Reply:
What would be the drawback of boxing it?

Reply:
> What would be the drawback of boxing it?

- Performance
- We have to decide whether we add the `Send` bound. In the current design, we get to delete the `YamuxLocal` stuff in `rust-libp2p` because the `Send` bound is inferred.

I don't feel strongly about either but it felt like a nice improvement as I went along. Once we get "impl Trait in type-alias" in Rust at least the boxing would go away.

Reply:
> > What would be the drawback of boxing it?
> 
>     * Performance

Is there any benchmark proving this? Is performance relevant when closing a connection?

> We have to decide whether we add the `Send` bound. In the current design, we get to delete the `YamuxLocal` stuff in `rust-libp2p` because the `Send` bound is inferred.

The infectious-send-when-boxing problem is reason enough to not box in my eyes :+1: